libraries/ESP8266WebServer/examples/FSBrowser/data/edit/index.htm

```diff
@@ -0,0 +1,1122 @@
+          if (xmlHttp.status == 200) {
+            editor.setValue(xmlHttp.responseText);
+            editor.clearSelection();
+            enableSaveDiscardBtns(false);
+          }
+          else tree.clearMainPanel();
+        }
+        function loadFile(path){
+          setLoading(true, "Loading '" + path + "'...");
+          xmlHttp = new XMLHttpRequest();
+          xmlHttp.onload = fileLoaded;
+          xmlHttp.open("GET", path, true);
+          xmlHttp.send(null);
+        }
+  
+        if (lang !== "plain") editor.getSession().setMode("ace/mode/"+lang);
+        editor.setTheme("ace/theme/"+theme);
+        editor.$blockScrolling = Infinity;
+        editor.getSession().setUseSoftTabs(true);
+        editor.getSession().setTabSize(2);
+        editor.setHighlightActiveLine(true);
+        editor.setShowPrintMargin(false);
+        editor.commands.addCommand({
+            name: 'save',
+            bindKey: {win: 'Ctrl-S',  mac: 'Command-S'},
+            exec: function(editor) {
+              editor.save();
+            },
+            readOnly: false
+        });
+        editor.commands.addCommand({
+            name: "showKeyboardShortcuts",
+            bindKey: {win: "Ctrl-Alt-h", mac: "Command-Alt-h"},
+            exec: function(editor) {
+                editor.showShortcuts();
+            }
+        })
+
+        editor.session.on('change', function(delta) {
+          enableSaveDiscardBtns(true);
+        });        
+
+        editor.loadUrl = function(path){
+          document.getElementById("pathInput").value = path;
+          enableSaveDiscardBtns(false);
+          file = path;
+          lang = getLangFromFilename(file);
+          type = "text/"+lang;
+          if (lang !== "plain") editor.getSession().setMode("ace/mode/"+lang);
+          loadFile(file);
+        }
+        
+        editor.save = function() {
+          enableSaveDiscardBtns(false);
+          postFile(file, editor.getValue()+"", type);
+        }
+        
+        editor.discard = function() {
+          editor.loadUrl(file);
+          enableSaveDiscardBtns(false);
+        }
+        
+        editor.showShortcuts = function() {
+          ace.config.loadModule("ace/ext/keybinding_menu", function(module) {
+            module.init(editor);
+            editor.showKeyboardShortcuts()
+          });
+        }
+
+        loadFile(file);        
+        
+        return editor;
+      }
+      
+      /////////////////////////
+      // MAIN ENTRY POINT
+      
+      function onBodyLoad(){
+        var vars = {};
+        var parts = window.location.href.replace(/[?&]+([^=&]+)=([^&]*)/gi, function(m,key,value) { vars[key] = value; });        
+        if (typeof ace != "undefined") {
+          var editor = createEditor("editor", vars.file, vars.lang, vars.theme);
+        }
+        tree = createTree("tree", editor);
+        createHeader("header", tree, editor);
+        refreshStatus();
+      };
+    </script>
```

Comment:
Does it make sense to move this whole thing into its own js?
It's considered best practice to try to maintain separation between html/js/css as far as is reasonable.

Reply:
See comment for the CSS.
I think this sketch is half an example of how to use the ESP8266WebServer lib (plus many other things, such as FileSystem calls), and half a real utility that could become a separate project on its own.
So it is a balance between education (which would tend to make things more explicit, reduce fancy cases, etc) and real-world usage.
I admit what motivated me to dig down inside the code is that I include that sketch in all my ESP projects as a way to debug and manipulate my filesystem contents.